### PR TITLE
gh-96348: Deprecate the 3-arg signature of coroutine.throw, generator.throw and agen.athrow

### PR DESCRIPTION
**Pull request by ofey404**
Add an deprecation warning, when `nargs > 1` in `Objects/genobject.c:gen_throw`.

- Also change corresponding docstrings.

* Issue: gh-96348

### CODE REVIEW

File: Objects/genobject.c [hunk 1]
```diff
@@ -559,6 +561,14 @@ gen_throw(PyGenObject *gen, PyObject *const *args, Py_ssize_t nargs)
     if (!_PyArg_CheckPositional("throw", nargs, 1, 3)) {
         return NULL;
     }
+    if (nargs > 1) {
+        if (PyErr_WarnEx(PyExc_DeprecationWarning,
+                            "the (type, val, tb) exception representation"
+                            "is deprecated, and may be removed in a future version of Python.",
```
Comment (gvanrossum) [inline Objects/genobject.c]:
```suggestion
                            "the (type, val, tb) exception representation is deprecated, "
                            "and may be removed in a future version of Python.",
```

**Review by gvanrossum — COMMENTED** [review]
One nit. Too bad about the failing tests, let's brainstorm what to do about them:

- We could just fix them, replacing `.throw(A, B, C)` with `.throw(B)`.
- We could add `filterwarnings()` calls to the offending tests (I don't recall how to do that but there are probable plenty of examples).
- We could report a `SilentDeprecationWarning`, then gradually fix the tests, and then later change it into a `DeprecationWarning`.
- We could first fix the tests, in batches, and then land this PR.

**Comment by gvanrossum** [comment]
> We could report a `SilentDeprecationWarning`, then gradually fix the tests, and then later change it into a `DeprecationWarning`.

Whoops, there's no such thing. Maybe it once existed, but it doesn't now, so ignore this option. :-)

**Comment by ofey404** [comment]
> One nit. Too bad about the failing tests, let's brainstorm what to do about them:

Let's estimate the problem size, by finding with regex `throw\(.*,.*,.*\)`

There are 51 results in 12 files.

- 3 file is under `Lib`
- 1 is in `Lib/lib2to3/fixes`
- The rest are tests.
- [51 results in 12 files.txt](https://github.com/python/cpython/files/9457682/51.results.in.12.files.txt)

----

> - We could just fix them, replacing .throw(A, B, C) with .throw(B).
> - We could first fix the tests, in batches, and then land this PR.

I prefer fixing them in this PR, for the fix need to be verified by raising an error. Otherwise the batch fix might miss something.

- However, the batch plan would be clearer, in Issue and PR history.

> We could add filterwarnings() calls to the offending tests (I don't recall how to do that but there are probable plenty of examples).

`filterwarnings()` call might be a safety valve, if I am not quite confident about how some tests work.

**Comment by gvanrossum** [comment]
Okay, make sure you consider [this](https://github.com/python/cpython/issues/96348#issuecomment-1232217457).

Also, some of the hits may be tests for this very feature, those should not be "fixed" but you may have to add a `with self.warns()` on those.

**Comment by ofey404** [comment]
I run `./python -m test` and fixed cases which are clear to be changed to `.throw(val)` form.

However, there are still some `DeprecationWarning`s need to be handle. As @iritkatriel said in [comment of 96348](https://github.com/python/cpython/issues/96348#issuecomment-1232507774), some of the tests need to stay.

## 1. Generator's doc test

I think we should filter the `DeprecationWarning` here, for it's testing exactly the deprecated API.

https://github.com/python/cpython/blob/0cd33e11fe6c77c423dbf3a7a9920daff012f006/Lib/test/test_generators.py#L2116-L2130


## 2. Lib `unittest.case`

It happens here:

https://github.com/python/cpython/blob/0cd33e11fe6c77c423dbf3a7a9920daff012f006/Lib/unittest/case.py#L236-L237

The output is like:

```plain text
/../cpython/Lib/unittest/case.py:237: DeprecationWarning: the (type, val, tb) exception representationis deprecated, and may be removed in a future version of Python.
  callable_obj(*args, **kwargs)
```

I am not sure what `callable_obj` does here, and maybe filtering this warning requires a change in standard library of python?

File: Objects/genobject.c [hunk 2]
```diff
@@ -559,6 +561,14 @@ gen_throw(PyGenObject *gen, PyObject *const *args, Py_ssize_t nargs)
     if (!_PyArg_CheckPositional("throw", nargs, 1, 3)) {
         return NULL;
     }
+    if (nargs > 1) {
+        if (PyErr_WarnEx(PyExc_DeprecationWarning,
+                            "the (type, val, tb) exception representation is deprecated, "
+                            "and may be removed in a future version of Python.",
```
Comment (iritkatriel) [inline Objects/genobject.c]: @gvanrossum "may be" or "will be"?

File: Objects/genobject.c [hunk 3]
```diff
@@ -418,7 +418,9 @@ PyDoc_STRVAR(throw_doc,
 throw(type[,value[,tb]])\n\
 \n\
 Raise exception in generator, return next yielded value or raise\n\
-StopIteration.");
+StopIteration.\n\
+the (type, val, tb) exception representation is deprecated, \n\
+and may be removed in a future version of Python.");
```
Comment (iritkatriel) [inline Objects/genobject.c:423]: The documentation needs to be updates in 
https://docs.python.org/3.12/reference/expressions.html#generator.throw

File: Misc/NEWS.d/next/Core and Builtins/2022-08-31-18-46-13.gh-issue-96348.xzCoTP.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+Deprecate ``gen.throw(typ, val, tb)``
```
Comment (iritkatriel) [inline Misc/NEWS.d/next/Core and Builtins/2022-08-31-18-46-13.gh-issue-96348.xzCoTP.rst]:
```suggestion
Emit a DeprecationWarning when :meth:`~generator.throw` or :meth:`~coroutine.throw` are called with more than one argument.
```

**Comment by iritkatriel** [comment]
> ```
> /../cpython/Lib/unittest/case.py:237: DeprecationWarning: the (type, val, tb) exception representationis deprecated, and may be removed in a future version of Python.
>   callable_obj(*args, **kwargs)
> ```
> 
> I am not sure what `callable_obj` does here, and maybe filtering this warning requires a change in standard library of python?

Look at the code in Lib/unittest/case.py. This is part of the assertRaises mechanism. The callable_obj and the args passed to it are defined by the test, so this should be dealt with in the particular test and not here in the test framework.

**Comment by iritkatriel** [comment]
> >   
> > I am not sure what `callable_obj` does here, and maybe filtering this warning requires a change in standard library of python?
> 
> 
> Look at the code in Lib/unittest/case.py. This is part of the assertRaises mechanism. The callable_obj and the args passed to it are defined by the test, so this should be dealt with in the particular test and not here in the test framework.


The test is this one. It's testing the types of the triplet, so we need to keep it and suppress the deprecation warning.

```
    def test_future_iter_throw(self):
        fut = self._new_future(loop=self.loop)
        fi = iter(fut)
        self.assertRaises(TypeError, fi.throw, 
                          Exception, Exception("elephant"), 32)
        self.assertRaises(TypeError, fi.throw,
                          Exception("elephant"), Exception("elephant"))
        self.assertRaises(TypeError, fi.throw, list)
```

File: Objects/genobject.c (hunk 2 above)
Comment (gvanrossum) [reply Objects/genobject.c]: Based on the experience with PEP 585 let’s keep may”.

**Comment by ofey404** [comment]
Okay, handled the remaining warnings.

Tested with those scripts:

```bash
./python -m test -j16 > test.log; cat test.log | grep "DeprecationWarning: the (type, val, tb) exception representation is deprecated"
# Come out with nothing

./python -Werror -m unittest -v test.test_asyncio.test_futures.PyFutureTests.test_future_iter_throw test.test_generators
# OK
```

Besides, should I squash all commits into one, when all the reviews are done?

**Comment by gvanrossum** [comment]
Don’t squash please! It makes review harder. We will squash upon merge.

File: Objects/genobject.c (hunk 3 above)
Comment (iritkatriel) [reply Objects/genobject.c:423]: The doc update is still missing.

**Review by kumaraditya303 — CHANGES_REQUESTED** [review]
This is a significant change, it should be properly documented in 3.12 What's New.

File: Objects/genobject.c (hunk 3 above)
Comment (ofey404) [reply Objects/genobject.c:423]: Added a deprecation line in `Doc/reference/expressions.rst`.

I refer to [Doc/library/ast.rst](https://github.com/python/cpython/blob/main/Doc/library/ast.rst). But I'm not sure how to mark a single signature as deprecated, like marking deprecated methods in `ast.rst`:

```plain text
      Methods :meth:`visit_Num`, :meth:`visit_Str`, :meth:`visit_Bytes`,
      :meth:`visit_NameConstant` and :meth:`visit_Ellipsis` are deprecated
      now and will not be called in future Python versions.  Add the
      :meth:`visit_Constant` method to handle all constant nodes.
```

File: Doc/reference/expressions.rst [hunk 5]
```diff
@@ -582,6 +582,11 @@ is already executing raises a :exc:`ValueError` exception.
    :attr:`~BaseException.__traceback__` attribute stored in *value* may
    be cleared.
 
+   .. deprecated:: 3.12
```
Comment (iritkatriel) [inline Doc/reference/expressions.rst]: I think this would just be 

`.. versionchanged:: 3.12`

File: Doc/reference/expressions.rst [hunk 6]
```diff
@@ -582,6 +582,11 @@ is already executing raises a :exc:`ValueError` exception.
    :attr:`~BaseException.__traceback__` attribute stored in *value* may
    be cleared.
 
+   .. versionchanged:: 3.12
+
+      The second signature \(type\[, value\[, traceback\]\]\) is deprecated now.
+      It is still available, but may be removed in a future version of Python.
```
Comment (iritkatriel) [inline Doc/reference/expressions.rst]:
```suggestion
      The second signature \(type\[, value\[, traceback\]\]\) is deprecated and
      may be removed in a future version of Python.
```

Comment (iritkatriel) [reply Doc/reference/expressions.rst]: There are a few other places in the docs (I used index search for throw):

https://docs.python.org/3.12/reference/datamodel.html?highlight=throw#coroutine.throw
https://docs.python.org/3.12/reference/expressions.html?highlight=throw#agen.athrow

And as @kumaraditya303 mentioned, we need an entry in 

https://github.com/python/cpython/blob/main/Doc/whatsnew/3.12.rst

**Comment by ofey404** [comment]
I have some questions about writing a What's New entry:

- Does Issue #89874 has a relevant PEP?
- Should we place this change in the **Optimizations** section or the **Deprecated** or **Pending Removal**?

**Comment by iritkatriel** [comment]
> I have some questions about writing a What's New entry:
> 
> * Does Issue [Simplify the interpreter's (type, val, tb) exception representation #89874](https://github.com/python/cpython/issues/89874) has a relevant PEP?

no, there is no pep. 

> * Should we place this change in the **Optimizations** section or the **Deprecated** or **Pending Removal**?

deprecated fits best (probably under the stdlib heading).

File: Doc/whatsnew/3.12.rst [hunk 7]
```diff
@@ -159,6 +159,8 @@ Deprecated
   and tailor them to your needs.
   (Contributed by Erlend E. Aasland in :gh:`90016`.)
 
+* The exception representation \(type\[, value\[, traceback\]\]\) is now deprecated. (:gh:`89874`.)
```
Comment (iritkatriel) [inline Doc/whatsnew/3.12.rst]: This PR doesn't deprecate the exception representation ``(type, value, traceback)`` in general. It only deprecates the throw() signature that takes this representation.

File: Doc/whatsnew/3.12.rst [hunk 8]
```diff
@@ -159,7 +159,7 @@ Deprecated
   and tailor them to your needs.
   (Contributed by Erlend E. Aasland in :gh:`90016`.)
 
-* The exception representation \(type\[, value\[, traceback\]\]\) is now deprecated. (:gh:`89874`.)
+* The throw\(type, value, traceback\) signature is now deprecated. (:gh:`89874`.)
```
Comment (iritkatriel) [inline Doc/whatsnew/3.12.rst]: The 3-arg signatures ``(type, value, traceback)`` of :meth:`~coroutine.throw` and :meth:`~generator.throw` are deprecated and may be removed in a future version of Python. Use the single-arg versions of these functions instead.

File: Doc/reference/expressions.rst [hunk 9]
```diff
@@ -750,6 +756,11 @@ which are used to control the execution of a generator function.
    raises a different exception, then when the awaitable is run that exception
    propagates to the caller of the awaitable.
 
+   .. versionchanged:: 3.12
+
+      The second signature \(type\[, value\[, traceback\]\]\) is deprecated and
+      may be removed in a future version of Python.
```
Comment (iritkatriel) [inline Doc/reference/expressions.rst:762]: I don't think this PR actually raises a deprecation error for this one.

File: Objects/genobject.c [hunk 10]
```diff
@@ -418,7 +418,9 @@ PyDoc_STRVAR(throw_doc,
 throw(type[,value[,tb]])\n\
 \n\
 Raise exception in generator, return next yielded value or raise\n\
-StopIteration.");
+StopIteration.\n\
+the (type, val, tb) exception representation is deprecated, \n\
```
Comment (iritkatriel) [inline Objects/genobject.c]:
```suggestion
the (type, val, tb) signature is deprecated, \n\
```

File: Objects/genobject.c (hunk 2 above)
Comment (iritkatriel) [inline Objects/genobject.c]:
```suggestion
                            "the (type, exc, tb) signature of throw() is deprecated, "
                            "use the single-arg signature instead.",
```

File: Objects/genobject.c [hunk 11]
```diff
@@ -1147,7 +1157,10 @@ PyDoc_STRVAR(coro_throw_doc,
 throw(type[,value[,traceback]])\n\
 \n\
 Raise exception in coroutine, return next iterated value or raise\n\
-StopIteration.");
+StopIteration.\n\
+the (type, val, tb) exception representation is deprecated, \n\
```
Comment (iritkatriel) [inline Objects/genobject.c]:
```suggestion
the (type, val, tb) signature is deprecated, \n\
```

File: Lib/test/test_generators.py [hunk 12]
```diff
@@ -2113,6 +2113,13 @@ def printsolution(self, x):
 >>> g.throw(ValueError("xyz"))  # value only
 caught ValueError (xyz)
 
+>>> import warnings
+>>> warnings.filterwarnings("ignore", category=DeprecationWarning)
+
+# Filter DeprecationWarning: the (type, val, tb) exception representation is deprecated,
+#                            and may be removed in a future version of Python.
+# Will re-enable it soon.
```
Comment (iritkatriel) [inline Lib/test/test_generators.py]:
```suggestion
# Filter DeprecationWarning: regarding the (type, val, tb) signature of throw().
# Deprecation warnings are re-enabled below.
```

**Comment by gvanrossum** [comment]
Is this nearing completion?

**Comment by ofey404** [comment]
> Is this nearing completion?

hmmmmm……There are some documentation consistency issue yet. And I’m not sure how to edit the what’s new entry.

I’ll do some research about it tonight.

**Comment by iritkatriel** [comment]
Could you also double check that agen.athrow is indeed raising the deprecation error?

File: Doc/whatsnew/3.12.rst [hunk 13]
```diff
@@ -159,7 +159,9 @@ Deprecated
   and tailor them to your needs.
   (Contributed by Erlend E. Aasland in :gh:`90016`.)
 
-* The throw\(type, value, traceback\) signature is now deprecated. (:gh:`89874`.)
+* The 3-arg signatures (type, value, traceback) of :meth:`~coroutine.throw`
+  and :meth:`~generator.throw` are deprecated and may be removed in a future
+  version of Python. Use the single-arg versions of these functions instead. (:gh:`89874`.)
```
Comment (iritkatriel) [inline Doc/whatsnew/3.12.rst]:
```suggestion
  version of Python. Use the single-arg versions of these functions instead.
  (Contributed by Ofey Chan in :gh:`89874`.)
```

Comment (iritkatriel) [reply Doc/whatsnew/3.12.rst]: You can also add yourself to https://github.com/python/cpython/blob/main/Misc/ACKS

**Comment by ofey404** [comment]
> Or you can make the code change so that it does raise.

Cool. I’ll do it tomorrow.

**Comment by iritkatriel** [comment]
> Add DeprecationWarning, doc and what's new to `agen.athrow`. To my surprise, there is no usage of `agen.athrow` in the test suite of python.

Could you add one while you're here?

> 
> And I found another occurance of the 3-arg signature:
> 
> https://github.com/python/cpython/blob/10e3d398c31cc1695752fc52bc6ca2ce9ef6237e/Objects/iterobject.c#L430-L432
> 
> It seems to be a wrapper for calling `athrow` in awaitable object, maybe we should update the doc here too?

Yes I think that doc should be updated as well. Good find.

**Comment by ofey404** [comment]
> Could you add one while you're here?

A new test catching DeprecationWarning from 3-arg `agen.athrow()`. ;)

File: Lib/test/test_asyncgen.py [hunk 14]
```diff
@@ -377,6 +378,13 @@ async def async_gen_wrapper():
 
         self.compare_generators(sync_gen_wrapper(), async_gen_wrapper())
 
+    def test_async_gen_3_arg_deprecation_warning(self):
+        async def gen():
+            yield 123
```
Comment (iritkatriel) [inline Lib/test/test_asyncgen.py:383]:
```suggestion
            yield 123
            
```

(PEP 8)

File: Doc/whatsnew/3.12.rst [hunk 15]
```diff
@@ -159,6 +159,11 @@ Deprecated
   and tailor them to your needs.
   (Contributed by Erlend E. Aasland in :gh:`90016`.)
 
+* The 3-arg signatures (type, value, traceback) of :meth:`~coroutine.throw`
+  , :meth:`~generator.throw` and :meth:`~agen.athrow` are deprecated and may be removed in a future
+  version of Python. Use the single-arg versions of these functions instead.
+  (Contributed by Ofey Chan in :gh:`89874`.)
```
Comment (iritkatriel) [inline Doc/whatsnew/3.12.rst]:
```suggestion
* The 3-arg signatures (type, value, traceback) of :meth:`~coroutine.throw`,
  :meth:`~generator.throw` and :meth:`~agen.athrow` are deprecated and
  may be removed in a future version of Python. Use the single-arg versions
  of these functions instead. (Contributed by Ofey Chan in :gh:`89874`.)
```

(changed only whitespace and wrapping).

**Comment by iritkatriel** [comment]
Thanks, looks good now. I made two formatting suggestions.

Could you also create an issue about the fact that there are no tests for agen.athrow (other than the new one you added here)?

**Comment by ofey404** [comment]
> Thanks, looks good now. I made two formatting suggestions.
> 
> Could you also create an issue about the fact that there are no tests for agen.athrow (other than the new one you added here)?

Sure. Besides, I don’t find formatting suggestions, are they new added or have been closed?

And should I squash all commits into one, or this procedure would be done by a core developer?

**Comment by iritkatriel** [comment]
Sorry I missed a click before. You should see them now.

Don’t squash the commits, we do it at the merge.

**Comment by ofey404** [comment]
> Could you also create an issue about the fact that there are no tests for agen.athrow (other than the new one you added here)?

I checked and there are tests for `agen.athrow`, my fault. There is just no usage of 3-arg arguments.

Tests like:
https://github.com/python/cpython/blob/817fa28f81eed43539fad2c8e696df954afa6ad7/Lib/test/test_asyncgen.py#L368-L376

**Comment by ofey404** [comment]
Besides, Azure Pipeline seems failing on whitespace of Lib/test/test_asyncgen.py.

I might run a formatter, but which one does cpython project use? I search the web but don't find the answer.

**Comment by iritkatriel** [comment]
> Besides, Azure Pipeline seems failing on whitespace of Lib/test/test_asyncgen.py.
> 
> I might run a formatter, but which one does cpython project use? I search the web but don't find the answer.

It's running the patchcheck.py script which is in the Tools directory. You can run that.

**Review by iritkatriel — APPROVED** [review]
Fix that patchcheck errors, otherwise LGTM.

File: Misc/NEWS.d/next/Core and Builtins/2022-08-31-18-46-13.gh-issue-96348.xzCoTP.rst [hunk 16]
```diff
@@ -0,0 +1 @@
+Emit a DeprecationWarning when :meth:`~generator.throw` or :meth:`~coroutine.throw` are called with more than one argument.
```
Comment (iritkatriel) [inline Misc/NEWS.d/next/Core and Builtins/2022-08-31-18-46-13.gh-issue-96348.xzCoTP.rst]:
```suggestion
Emit a DeprecationWarning when :meth:`~generator.throw`, :meth:`~coroutine.throw` or :meth:`~agen.athrow`
are called with more than one argument.
```

File: Lib/test/test_asyncgen.py [hunk 17]
```diff
@@ -377,6 +378,14 @@ async def async_gen_wrapper():
...
+
+        with warnings.catch_warnings():
```
Comment (kumaraditya303) [inline Lib/test/test_asyncgen.py]: This should test that `DeprecationWarning` is emitted rather than ignoring it. Same for all other places.

**Review by kumaraditya303 — CHANGES_REQUESTED** [review]

File: Lib/test/test_asyncgen.py (hunk 17 above)
Comment (iritkatriel) [reply Lib/test/test_asyncgen.py]: Good point.

(One test for each of the 3 deprecated APIs should check that the deprecation warning is raised. Where the point of the test is to check something else, the deprecation warning can be suppressed).

File: Lib/test/test_asyncio/test_futures.py [hunk 18]
```diff
@@ -619,10 +620,14 @@ def test_future_stop_iteration_args(self):
     def test_future_iter_throw(self):
         fut = self._new_future(loop=self.loop)
         fi = iter(fut)
-        self.assertRaises(TypeError, fi.throw,
-                          Exception, Exception("elephant"), 32)
-        self.assertRaises(TypeError, fi.throw,
-                          Exception("elephant"), Exception("elephant"))
+        with self.assertWarns(DeprecationWarning):
+            fi.throw(Exception, Exception("zebra"), None)
```
Comment (iritkatriel) [inline Lib/test/test_asyncio/test_futures.py]:
```suggestion
            self.assertRaises(Exception, fi.throw, Exception, Exception("zebra"), None)
```

**Comment by iritkatriel** [comment]
It seems that my tests exposed some problem - there are 3 versions of futures, and only one of them is emitting the deprecation warning?

**Comment by iritkatriel** [comment]
Looking at the code, if futures._CFuture is fixed (made to raise the deprecation warning) then that will fix CSubFuture as well.

**Comment by ofey404** [comment]
> It seems that my tests exposed some problem - there are 3 versions of futures, and only one of them is emitting the deprecation warning?

The warning is added.

----

But I still feel a little dizzy about it - How do you know that we should add a test in test_futures.py? Is python future implemented by coroutine object or async generator?

- My guess: According to the doc, so test_future.py is testing the high-level interface of coroutine object, from that we know that we should add a test on `iter(future).throw`.

>The [concurrent.futures](https://docs.python.org/3/library/concurrent.futures.html#module-concurrent.futures) module provides a high-level interface for asynchronously executing callables.
> [concurrent.futures — Launching parallel tasks](https://docs.python.org/3/library/concurrent.futures.html)

**Comment by iritkatriel** [comment]
@kumaraditya303 How does it look now? (I'm a little dizzy by now as well).

**Review by kumaraditya303 — APPROVED** [review]
LGTM

**Comment by kumaraditya303** [comment]
You only need to add deprecation warning in `_CFuture` since it is implemented in C. The pure Python versions uses regular generators so they will automatically emit warning since warning is added to `generator.throw`.

https://github.com/python/cpython/blob/9a11ed8e50492d327e4de0a8f3a473e788b14a6f/Lib/asyncio/futures.py#L290-L298

File: Doc/reference/datamodel.rst [hunk 19]
```diff
@@ -2996,6 +2996,11 @@ generators, coroutines do not directly support iteration.
    above.  If the exception is not caught in the coroutine, it propagates
    back to the caller.
 
+   .. versionchanged:: 3.12
+
+      The second signature \(type\[, value\[, traceback\]\]\) is deprecated and
```
Comment (merwok) [inline Doc/reference/datamodel.rst:3001]: For future PRs, not that code should generally be marked up like

```
      The second signature `(type[, value[, traceback]])` is deprecated and
```

**Pull request merged**